### DISCUSSION TITLE
error TS2536: Type 'key' cannot be used to index type 'Store<string, S, G, {}>

### DISCUSSION THREAD

**Discussion by coader** [vuejs/pinia#2605] [Help and Questions] [answered]
### Reproduction

error TS2536:

### Steps to reproduce the bug

- error TS2536: Type 'key' cannot be used to index type 'Store<string, S, G, {}>'.

354     [key in Keys]: () => Store<string, S, G, {}>[key];
                             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 1 error in ../../node_modules/.pnpm/pinia@2.1.7_typescript@5.4.2_vue@3.4.21/node_modules/pinia/dist/pinia.d.ts:354

### Expected behavior

it's ok yesterday, I  don't know why and which package upgrade bring this

### Actual behavior

- error TS2536: Type 'key' cannot be used to index type 'Store<string, S, G, {}>'.

### Additional information

_No response_

**Comment by posva** [comment]
You will need to provide the code that triggers this error and more information if you want to get help 😉

**Reply by coader** [reply]
it's only throw error when build with type check
any code use pinia and typescript 5.4 can't build successfully

**Reply by coader** [reply]
`export declare type _MapStateReturn<S extends StateTree, G extends _GettersTree<S>, Keys extends keyof S | keyof G = keyof S | keyof G> = {
    [key in Keys]: () => Store<string, S, G, {}>[key];
};`


`error TS2536: Type 'key' cannot be used to index type 'Store<string, S, G, {}>'.

354     [key in Keys]: () => Store<string, S, G, {}>[key];
                             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Comment by coader** [comment]
![image](https://github.com/vuejs/pinia/assets/3700498/4edd3848-f6ba-457f-b407-60d370660954)

not need any code, just error when build and the error is in d.ts

**Comment by coader** [comment]
it looks like that because typescript upgraded to 5.4

**Comment by itsmnthn** [comment]
Correct faced same thing, especially for getters with objects

Updating to `typescript@5.4.x` caused this, reverting back to `5.3.x` was the solution for me

Reproduction
https://stackblitz.com/edit/nuxt-starter-tf7qkb?file=pages%2Findex.vue

![Screenshot 2024-03-23 at 3 59 11 PM](https://github.com/vuejs/pinia/assets/25129674/2295a249-7468-40a6-beef-31a3b2fd9f8e)

![Screenshot 2024-03-23 at 3 59 32 PM](https://github.com/vuejs/pinia/assets/25129674/eb36e213-a25c-41aa-a44f-8555424a480a)

**Reply by bodograumann** [reply]
Same here. Curiously I only saw it in a library-project, not in the vue-app project :-/

**Comment by coader** [comment]
already discard to use pinia
as I am not use SSR, so global reactive object is works fine for me

**Comment by Zellement** [comment]
Downgrading Typescript to `5.3.3` also fixed this for me.

```
 "typescript": "5.3.3"
```

**Comment by dten** [comment]
i made a repro here that shows upgrading to ts 5.4.5 causes pinia not to type check

https://github.com/dten/pinia_ts_repro/actions/runs/8710652508/job/23893007875

**Comment by sTeVeNNs** [comment, marked as answer]
When I added the following line to the ts compilerOptions, the error fixed：
`"skipLibCheck": true`
see: https://github.com/vuejs/vue-cli/pull/5688

**Reply by dten** [reply]
That just ignores the error rather than fixes

**Reply by sTeVeNNs** [reply]
Yeah, you're right.

**Reply by posva** [reply]
Until very recently, the codebase was failing with recent versions of TS, hence the error you see if you type check the library code. In applications, it's better not to type-check libraries for the reasons mentioned in the PR. In the case of Pinia, the one that applied was:

> Both TypeScript and Vue minor version updates, though as widely tested as they can go, still have chances to break existing type definitions. Users' projects should not break because of code they can't control.

If this is still unclear and want to know more, I recommend you to read the references issues in the PR.

**Comment by loyoi** [comment]
gpt 给的答案

`export declare type _MapStateReturn<S extends StateTree, G extends _GettersTree<S>, Keys extends keyof S | keyof G = keyof S | keyof G> = {
    [key in Keys]: () => Store<string, S, G, {}>[keyof Store<string, S, G, {}>];
};`

**Comment by Penumbra69** [comment]
Are there any solutions here other than downgrade typescript a whole minor revision?
I see GPT's solution above, but .. I'm not gonna monkey patch pinia to get it working in 5.4.5 am I?!

**Reply by reinaldomendes** [reply]
I patched the pinia with pnpm patch and added a @ts-expect-error on the line. 

I'm waiting for the fixes.

```diff
diff --git a/dist/pinia.d.ts b/dist/pinia.d.ts
index 7b5764669f13b99c3e8d58acbeb37eb4bb3e3b32..0a86fca9be2e353906ea8ad62c158b05559d88ca 100644
--- a/dist/pinia.d.ts
+++ b/dist/pinia.d.ts
@@ -351,6 +351,7 @@ export declare type _MapStateObjectReturn<Id extends string, S extends StateTree
  * For internal use **only**
  */
 export declare type _MapStateReturn<S extends StateTree, G extends _GettersTree<S>, Keys extends keyof S | keyof G = keyof S | keyof G> = {
+    // @ts-expect-error https://github.com/vuejs/pinia/discussions/2605
     [key in Keys]: () => Store<string, S, G, {}>[key];
 };
